File: src/ru/edu/vtisov/stack/MinStack.java
package ru.edu.vtisov.stack;

import java.util.LinkedList;

// https://leetcode.com/problems/min-stack/

public class MinStack {

    private final LinkedList<Integer> stack;
    private final LinkedList<Integer> min;
    
    public MinStack() {
        this.stack = new LinkedList<>();
        this.min = new LinkedList<>(); 
    }
    
    public void push(int val) {
        stack.add(val);
        min.add(Math.min(val, min.size() > 0 ? min.getLast() : val));
    }
    
    public void pop() {
        min.removeLast();
        stack.removeLast();
    }
    
    public int top() {
        return stack.getLast();
    }
    
    public int getMin() {
        return min.getLast();
    }
}

File: src/ru/edu/vtisov/stack/min-stack.go
package stack

//https://leetcode.com/problems/min-stack/description/

type MinStack struct {
	stack [][]int
}

func Constructor() MinStack {
	return MinStack{stack: make([][]int, 0)}
}

func (this *MinStack) Push(value int) {
	if len(this.stack) == 0 {
		this.stack = append(this.stack, []int{value, value})
		return
	}
	currentMin := min(this.stack[len(this.stack)-1][1], value)
	this.stack = append(this.stack, []int{value, currentMin})
}

func (this *MinStack) Pop() {
	this.stack = this.stack[:len(this.stack)-1]
}

func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1][0]
}

func (this *MinStack) GetMin() int {
	return this.stack[len(this.stack)-1][1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(value);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
